packages/web3/src/contract/contract.ts

```diff
@@ -821,19 +862,22 @@ export class Project {
     projectRootDir = '.',
     contractsRootDir = Project.DEFAULT_CONTRACTS_DIR,
     artifactsRootDir = Project.DEFAULT_ARTIFACTS_DIR,
-    defaultFullNodeVersion: string | undefined = undefined
+    defaultFullNodeVersion: string | undefined = undefined,
+    skipSaveArtifacts = false
   ): Promise<void> {
     const provider = getCurrentNodeProvider()
     const fullNodeVersion = defaultFullNodeVersion ?? (await provider.infos.getInfosVersion()).version
     const sourceFiles = await Project.loadSourceFiles(projectRootDir, contractsRootDir)
     const { errorOnWarnings, ...nodeCompilerOptions } = { ...DEFAULT_COMPILER_OPTIONS, ...compilerOptionsPartial }
     const projectArtifact = await ProjectArtifact.from(projectRootDir)
+    const changedSources = projectArtifact?.getChangedSources(sourceFiles) ?? sourceFiles
     if (
       projectArtifact === undefined ||
-      projectArtifact.needToReCompile(nodeCompilerOptions, sourceFiles, fullNodeVersion)
+      projectArtifact.needToReCompile(nodeCompilerOptions, fullNodeVersion) ||
+      changedSources.length > 0
```

Comment:
Q: Shall we check `skipSaveArtifacts` here? I suppose when `skipSaveArtifacts` is false and project is deployed on mainnet, we still want to potentially regenerate the code.

Reply:
If the project is deployed on mainnet, then the `skipSaveArtifacts` should be true, we have already checked this in the CLI.

If the `skipSaveArtifacts` is true and the contract code is unchanged, then we won't regenerate bytecode, but If the contract code is changed, we will regenerate bytecode.

Reply:
I see, thanks!